central/productusage/injector/injector.go

```diff
@@ -0,0 +1,94 @@
+package injector
+
+import (
+	"context"
+	"time"
+
+	datastore "github.com/stackrox/rox/central/productusage/datastore/securedunits"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+const period = 1 * time.Hour
+
+var (
+	metricsWriter = sac.AllowFixedScopes(
+		sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+		sac.ResourceScopeKeys(resources.Administration))
+
+	once     sync.Once
+	injector Injector
+	log      = logging.LoggerForModule()
+)
+
+func (i *injectorImpl) gather(ctx context.Context) {
+	newMetrics, err := i.ds.AggregateAndFlush(ctx)
+	if err != nil {
+		log.Debug("Failed to get and flush the aggregated usage metrics: ", err)
+		return
+	}
+	ctx = sac.WithGlobalAccessScopeChecker(ctx, metricsWriter)
+	if err := i.ds.Upsert(ctx, newMetrics); err != nil {
+		log.Debug("Failed to store a usage snapshot: ", err)
+	}
+}
+
+func (i *injectorImpl) gatherLoop() {
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+	ctx, cancel := context.WithCancel(context.Background())
+	i.gather(ctx)
+	for {
+		select {
+		case <-ticker.C:
+			i.gather(ctx)
+		case <-i.stop.Done():
+			cancel()
```

Comment:
Perhaps I'm forgetting my golang concurrency lessons again, but wouldn't `cancel` achieve nothing here since `case <-i.stop.Done()` won't run in parallel with `case <-ticker.C`?

Reply:
Good catch. It should `go i.gather(ctx)` to be cancellable.